File: algorithms/javascript/Ip.js

var recursion = function(s,l,c=[]){
  if (s.length === 0){
    if (c.length === 4){
      l.push(c);
      return true;
    }
    return false;
  }
  for(var n=1; n<4; ++n){
    p = c.slice();
    v = s.slice(0,n);
    if ((v.startsWith('0') && v.length > 1) || v > 255){
      continue;
    }
    p.push(v);
    if(recursion(s.slice(n),l,p)){
      break;
    }
  }
}

var initiator = function(s){
  l = [];
  recursion(s,l);
  for(var i=0; i<l.length; ++i){
    l[i] = l[i].join('.');
  }
  return l;
}

var ip = "255255255255";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "25525511255";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "25551125";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "2552";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "0255255255";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "0000";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "001100";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "001900";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "001994482909400";
console.log(ip + " -> [" + initiator(ip)+"]");
ip = "011";
console.log(ip + " -> [" + initiator(ip)+"]");

File: algorithms/python/Ip.py
from copy import deepcopy

def recursion(s,l,c=[]):
    if len(s) == 0:
        if len(c) == 4:
            l.append(c)
            return True
        return False
    for n in range(1,4):
        p = deepcopy(c)
        v = s[0:n]
        if (v.startswith('0') and len(v) > 1) or int(v) > 255:
            continue
        p.append(v)
        if recursion(s[n:], l, p):
            break

def initiator(s):
    l = []
    recursion(s,l)
    return ['.'.join(p) for p in l]

ip = "255255255255"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "25525511255"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "25551125"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "2552"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "0255255255"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "0000"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "001100"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "001900"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "001994482909400"
print('{} -> {}'.format(ip, initiator(ip)))
ip = "011"
print('{} -> {}'.format(ip, initiator(ip)))
